Stack to construct

greglucas opened the pull request:
# Change Summary

## Overview

This is a major refactor of our AWS infrastructure. It is best practice to focus on lifecycle for Stacks and deployment updates to avoid circular dependencies. We were running into a few circular dependencies because we had Stack A depending on Stack B, which would then want to try and update Stack A again. Putting everything into the same deployment unit makes sure that these updates get processed and deployed in the correct order.

There is a lot of thrash just changing names from "abc_stack" to "abc_construct", might be advantageous to avoid the "_specific_name" in the future...

We have 3 primary stacks now, with our previous fine-grained Stacks turning into Constructs.
1. Networking Stack (sets up VPC)
2. SDC Stack (sets up all resources that the SDC uses/needs)
3. I-ALiRT Stack (sets up all resources for the I-ALiRT processing)

The easiest way to review this might be to step through by commits. I tried to break the process up into small incremental commits. Although there were two other PRs that were merged during this process and it was easier to merge them in than rebase through all the commits.

### Testing

I have deployed this all into a separate dev account and it deploys successfully and works as expected from an initial cut at depositing a file and querying the DB for that file.

greglucas committed:
MNT: Switch to a single SDC Stack

Rather than many small stacks, this consolidates into a single stack
to avoid circular dependencies and make deployments easier.

greglucas committed:
TST: Update the tests for Stack -> Construct conversion

greglucas committed:
MNT: Rename Stack -> Construct for all classes

Break out the stacks into Networking, SDC, and I-ALiRT

greglucas committed:
Merge remote-tracking branch 'upstream/dev' into stack-to-construct

Additionally, refactor the lambda layer into a construct and re-use that
and the code throughout the SDC Stack.

greglucas committed:
MNT: Move security groups into the Resources that own them

greglucas committed:
MNT: Cleanup unused rds security group ingress

greglucas committed:
MNT/FIX: Cleanup resources and fix issues during deployment

This is an update based on test deployments that fixes a few issues
that came up and adds comments about what was needed.

greglucas committed:
Merge ialirt-ingest-lambda into stack-to-construct

greglucas committed:
FIX: Only run CreateSchemaConstruct on RDS Creation

laspsandoval commented on sds_data_manager/utils/stackbuilder.py:
```diff
@@ -3,30 +3,30 @@
 from pathlib import Path
```
Minor comment about naming. Keep uniform conventions. For example we have:
IAlirtBucketConstruct
IalirtProcessingPrimary

So keep or remove Construct part of the name for the Constructs.

greglucas replied:
Yeah, this was a bit all over the place, and I didn't do a good job of consolidating. My preference would be to drop the `_construct` everywhere, see this comment:
> There is a lot of thrash just changing names from "abc_stack" to "abc_construct", might be advantageous to avoid the "_specific_name" in the future...

If you all don't mind the churn in this single PR I can do it now, but I was trying to make this a little more manageable to review.

laspsandoval replied:
No worries to do it in another PR.

laspsandoval commented on sds_data_manager/utils/stackbuilder.py:
```diff
@@ -262,7 +269,7 @@ def build_backup(scope: App, env: Environment, source_account: str):
     # This is the S3 bucket used by upload_api_lambda
-    backup_bucket_stack.BackupBucket(
+    backup_bucket_construct.BackupBucket(
         scope,
```
For all the other constructs the scope is equal to the name of the stack. Why not this one?

greglucas replied:
Good catch, this needs a Stack at some point, so I added one above it here to follow the same style as the others.

laspsandoval commented on sds_data_manager/utils/stackbuilder.py:
```diff
@@ -47,112 +47,135 @@ def build_sds(
+    lambda_code_directory = Path(__file__).parent.parent / "lambda_code"
+
+    lambda_code = lambda_.Code.from_asset(str(lambda_code_directory))
+    db_lambda_layer = lambda_layer_construct.LambdaLayerConstruct(
```
Why do we only have a single lambda layer and it's named for the database?

greglucas replied:
We want to build the dependencies only once (@tech3371 's previous PR), and attach that to all the Lambda Functions that need it. Here this is the database requirements layer. For instance, if you add "imap_processing" as a requirement for the SPICE kernel work, that could be a separate layer and only attached to the Lambda Functions that need that specific piece.

laspsandoval replied:
Thanks for the explanation. I'm still getting intuition on how this works.

laspsandoval reviewed (commented):
I left a few comments. I'll take some more time to try it out. Thanks!

greglucas committed:
FIX: Add stack to the backup bucket application

laspsandoval reviewed (commented):
I thought about this over the weekend. For I-ALiRT the ECR needs to be deployed before anything else. Afterwards an image needs to be pushed to it. And then the ECS can be deployed. If you deploy the ECS without having an image in the ECR then it times out. Probably everything worked for you since we already had an ECR and image in the dev account so no update was required.

I know there is probably a better way to do this, but for now that is how it works. So that could cause issues with the way you have currently set things up.

tech3371 commented on sds_data_manager/constructs/data_bucket_construct.py:
```diff
@@ -1,14 +1,14 @@
 """Configure the data bucket stack."""
```
```suggestion
"""Configure the data bucket construct."""
```

tech3371 commented on sds_data_manager/constructs/database_construct.py:
```diff
@@ -0,0 +1,156 @@
+"""Configure the database stack."""
```
```suggestion
"""Configure the database construct."""
```

tech3371 commented on sds_data_manager/constructs/domain_construct.py:
```diff
@@ -1,12 +1,11 @@
 """Configure the domain stack."""
```
```suggestion
"""Configure the domain construct."""
```

tech3371 commented on sds_data_manager/constructs/efs_construct.py:
```diff
@@ -1,10 +1,8 @@
 """Configure the EFS stack."""
```
```suggestion
"""Configure the EFS construct."""
```

tech3371 commented on sds_data_manager/constructs/ialirt_bucket_construct.py:
```diff
@@ -1,13 +1,13 @@
 """Configure the IAlirt bucket stack."""
```
```suggestion
"""Configure the IAlirt bucket construct."""
```

tech3371 commented on sds_data_manager/constructs/indexer_lambda_construct.py:
```diff
@@ -1,9 +1,6 @@
 """Configure the indexer lambda stack."""
```
```suggestion
"""Configure the indexer lambda construct."""
```

tech3371 commented on sds_data_manager/constructs/instrument_lambdas.py:
```diff
@@ -93,7 +86,7 @@ def __init__(
             security_groups=[rds_security_group],
             allow_public_subnet=True,
-            layers=batch_starter_layers,
+            layers=layers,
```
good catch!

tech3371 commented on sds_data_manager/constructs/lambda_layer_construct.py:
```diff
@@ -46,13 +45,9 @@ def __init__(
         )
 
-        layer = lambda_.LayerVersion(
+        self.layer = lambda_.LayerVersion(
```
I am glad construct removes output and import steps for layer and EFS.

greglucas committed:
MNT: Skip I-ALiRT Stack deployments

tech3371 commented on sds_data_manager/constructs/sds_api_manager_construct.py:
```diff
@@ -1,24 +1,22 @@
 """Configure the SDS API Manager stack."""
```
```suggestion
"""Configure the SDS API Manager construct."""
```

tech3371 commented on sds_data_manager/lambda_code/SDSCode/create_schema.py:
```diff
@@ -13,56 +15,20 @@
-        send_response(event, context, "FAILED")
+    # NOTE: If we run this when trying to delete the stack, it fails because
+    #       the RDS isn't available, so skip it here.
+    if event.get("RequestType") != "Create":
```
nice find!

greglucas commented:
Per the conversation earlier today, I have removed the I-ALiRT stack from the deployment in the most recent commit with a note to review and add it back in later once the updates have been made.

tech3371 reviewed (approved):
Looks great to me! Thank you for doing this!

greglucas committed:
DOC/MNT: Remove stack from docstrings

greglucas committed:
MNT: Remove manual security group from Lambda Function

greglucas committed:
ENH: Add Route53 domain name delegation and authorization

This adds the construct for delegating domain name authorization
to the root account through a URL approval managed by a lambda function.

bmcclellan-cu reviewed (approved):
Looks good to me! I didn't test the deployment, but if you ran this code today then I'd say it's good.

greglucas committed:
FIX: Add us-east-1 environment and root certificate

greglucas commented:
> Looks good to me! I didn't test the deployment, but if you ran this code today then I'd say it's good.

@bmcclellan-cu, I have just pushed up the latest changes that I deployed from today. The most recent couple of commits include the work from Jackson for the HostedZone certificate.

tech3371 reviewed (approved):
We have verified that it works as it should. So looks good to me!
